src/librustc/infer/opaque_types/mod.rs

```diff
@@ -722,11 +722,11 @@ where
             ty::Closure(def_id, ref substs) => {
                 // Skip lifetime parameters of the enclosing item(s)
 
-                for upvar_ty in substs.upvar_tys(def_id, self.tcx) {
+                for upvar_ty in substs.closure().upvar_tys(def_id, self.tcx) {
                     upvar_ty.visit_with(self);
                 }
 
-                substs.closure_sig_ty(def_id, self.tcx).visit_with(self);
+                substs.closure().closure_sig_ty(def_id, self.tcx).visit_with(self);
```

Comment:
Maybe we should rename the `closure_` methods to not have that prefix? So this would be `substs.closure().sig_ty(...)`?